### PR TITLE
#22662 Partial Function for onSignal

### LINKED ISSUE #22662: use PartialFunction for signals
See discussion in https://github.com/akka/akka-meta/issues/46

Also, see if we can improve the syntax for the two parameters in `Stateful`, Signal trait should not be sealed

### PR DESCRIPTION
**Pull request by 2m**
This is only refactoring. I did not add any logic for making sure [this](https://github.com/akka/akka/issues/22662#issuecomment-290657402) still works. All tests pass, so either that is handled already or there are missing tests.

Fixes #22662

### CODE REVIEW

File: akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala [hunk 1]
```diff
@@ -335,13 +335,21 @@ object Actor {
...
-    onSignal:  (ActorContext[T], Signal) ⇒ Behavior[T] = Behavior.unhandledSignal.asInstanceOf[(ActorContext[T], Signal) ⇒ Behavior[T]])
+    onSignal:  PartialFunction[(ActorContext[T], Signal), Behavior[T]] = Behavior.unhandledSignal.asInstanceOf[PartialFunction[(ActorContext[T], Signal), Behavior[T]]])
     extends ExtensibleBehavior[T] {
     override def management(ctx: AC[T], msg: Signal): Behavior[T] = onSignal(ctx, msg)
     override def message(ctx: AC[T], msg: T) = onMessage(ctx, msg)
     override def toString = s"Immutable(${LineNumbers(onMessage)})"
+
+    def onSignal(onSignal: PartialFunction[(ActorContext[T], Signal), Behavior[T]]): Immutable[T] =
+      new Immutable(onMessage, onSignal)
+  }
```
Comment (patriknw) [inline akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala]: change `MutableBehavior.onSignal`  also, but that should be something like:
```scala
def onSignal: PartialFunction[Signal, Behavior[T]] = PartialFunction.empty
```

File: akka-typed-tests/src/test/scala/akka/typed/scaladsl/adapter/AdapterSpec.scala [hunk 2]
```diff
@@ -27,41 +27,43 @@ object AdapterSpec {
...
-          Same
-        case "watch" ⇒
-          ctx.watch(ref)
-          Same
-        case "supervise-stop" ⇒
-          val child = ctx.actorOf(untyped1)
-          ctx.watch(child)
-          child ! ThrowIt3
-          child.tell("ping", ctx.self.toUntyped)
+    Immutable[String] {
```
Comment (patriknw) [inline akka-typed-tests/src/test/scala/akka/typed/scaladsl/adapter/AdapterSpec.scala]: it's a bit sad that the type annotation is needed when `onSignal` is used, but the other syntax was not beutiful

**Review by patriknw — CHANGES_REQUESTED** [review]
I think this is looking good, some comments...

File: akka-typed-tests/src/test/scala/akka/typed/scaladsl/adapter/AdapterSpec.scala (hunk 2 above)
Comment (johanandren) [inline akka-typed-tests/src/test/scala/akka/typed/scaladsl/adapter/AdapterSpec.scala]: I don't get why that changes by calling a method on the returned object?

Comment (patriknw) [inline akka-typed-tests/src/test/scala/akka/typed/scaladsl/adapter/AdapterSpec.scala]: some type inference is lost, but I don't know exactly why.
javac had same problem when I tried same thing in javadsl (and there this was a showstopper, see my other PR)

**Review by patriknw — APPROVED** [review]
LGTM, let's merge this first and I will rebase my PRs

File: akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala (hunk 1 above)
Comment (johanandren) [inline akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala]: What about the Java API signal handling?

**Review by johanandren — APPROVED** [review]
LGTM

File: akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala (hunk 1 above)
Comment (patriknw) [inline akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala]: For the immutable see my comment here: https://github.com/akka/akka/pull/22798#discussion_r113715029

For the mutable we might want to do something when adding the builders, cc @raboof

**Pull request merged**